AMDGPU/GlobalISel: RegBankLegalize rules for load

petar-avramovic opened the pull request:
Add IDs for bit width that cover multiple LLTs: B32 B64 etc.
"Predicate" wrapper class for bool predicate functions used to
write pretty rules. Predicates can be combined using &&, || and !.
Lowering for splitting and widening loads.
Write rules for loads to not change existing mir tests from old
regbankselect.

petar-avramovic commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112882?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#112866** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112866?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#112882** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112882?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#112864** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112864?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#112865](https://github.com/llvm/llvm-project/pull/112865) <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112865?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#112863** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112863?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#112862** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112862?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @petar-avramovic and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeHelper.cpp:
```diff
@@ -37,6 +37,97 @@ bool RegBankLegalizeHelper::findRuleAndApplyMapping(MachineInstr &MI) {
+    if (ByteOffset == 0) {
+      BasePtrPlusOffsetReg = BasePtrReg;
+    } else {
+      BasePtrPlusOffsetReg = MRI.createVirtualRegister({PtrRB, PtrTy});
+      Register OffsetReg = MRI.createVirtualRegister({PtrRB, OffsetTy});
+      B.buildConstant(OffsetReg, ByteOffset);
+      B.buildPtrAdd(BasePtrPlusOffsetReg, BasePtrReg, OffsetReg);
```
Would be nice if we could fold createVirtualRegister with bank into the build* methods

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURBLegalizeRules.cpp:
```diff
@@ -293,7 +506,87 @@ RegBankLegalizeRules::RegBankLegalizeRules(const GCNSubtarget &_ST,
+  });
+
+  Predicate isConst([](const MachineInstr &MI) -> bool {
+    // This is wierd. Can AS in MMO be different then AS on pointer?
```
Typo wierd.

Yes, the verifier doesn't enforce this (although maybe it should). In the cases where the addrspace comes from an underlying IR value, it definitely can mismatch

petar-avramovic commented:
ping

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -233,6 +327,126 @@ RegBankLegalizeRules::getRulesForOpc(MachineInstr &MI) const {
+    //   A == true jump to B
+    //   A == false jump to end or Y, result is A(false) or Y
+    // (A || B || ... X) or (A || B || ... X) && Y
+    //   A == true jump to end or Y, result is B(true) or Y
```
```suggestion
    //   A == true jump to end or Y, result is A(true) or Y
```

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -233,6 +327,126 @@ RegBankLegalizeRules::getRulesForOpc(MachineInstr &MI) const {
+    unsigned FJumpOffset;
+  };
+
+  SmallVector<Elt, 8> Expression;
```
Make this and `Elt` private. `Elt` is quite subtle and should be hidden as an implementation detail.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -233,6 +327,126 @@ RegBankLegalizeRules::getRulesForOpc(MachineInstr &MI) const {
+    Expression.push_back({Pred, false, 1, 1});
+  };
+
+  Predicate(SmallVectorImpl<Elt> &Expr) { Expression.swap(Expr); };
```
I find having a constructor that is destructive on its argument in this way to be quite surprising. It is against good patterns in modern C++. Better have it take an rvalue reference (&&-reference). Then its callers have to explicitly `std::move`, but that makes it more obvious what happens at the call sites.

Also, this constructor should be private just like `Elt`.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -233,6 +327,126 @@ RegBankLegalizeRules::getRulesForOpc(MachineInstr &MI) const {
+    return Result;
+  };
+
+  Predicate operator!() {
```
Should be a const method.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -233,6 +327,126 @@ RegBankLegalizeRules::getRulesForOpc(MachineInstr &MI) const {
+    return Predicate(NegExpression);
+  };
+
+  Predicate operator&&(Predicate &RHS) {
```
Similar to the constructor above, the baseline of this should be a const method that accepts a const reference as an argument.

Do you have a good argument for the rvalue-argument overload below? If not, please just remove it.

The same applies to `operator||`.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -290,7 +504,86 @@ RegBankLegalizeRules::RegBankLegalizeRules(const GCNSubtarget &_ST,
       .Any({{DivS64, S32}, {{Vgpr64}, {Vgpr32}, Ext32To64}});
 
-  addRulesForGOpcs({G_LOAD}).Any({{DivS32, DivP1}, {{Vgpr32}, {VgprP1}}});
+  bool hasUnAlignedLoads = ST->getGeneration() >= AMDGPUSubtarget::GFX12;
```
"unaligned" is a single word:
```suggestion
  bool hasUnalignedLoads = ST->getGeneration() >= AMDGPUSubtarget::GFX12;
```

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -116,6 +193,50 @@ void RegBankLegalizeHelper::lower(MachineInstr &MI,
+      } else {
+        B128 = LLT::scalar(128);
+      }
+      if (Size / 128 == 2)
+        splitLoad(MI, {B128, B128});
+      if (Size / 128 == 4)
+        splitLoad(MI, {B128, B128, B128, B128});
```
I would expect this to be an else-if chain with an `else llvm_unreachable` at the end.

nhaehnle commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeRules.cpp:
```diff
@@ -290,7 +504,86 @@ RegBankLegalizeRules::RegBankLegalizeRules(const GCNSubtarget &_ST,
+           (MemSize == 8 && MMO->getAlign() >= Align(1));
+  });
+
+  auto isUL = !isAtomicMMO && isUniMMO && (isConst || !isVolatileMMO) &&
+              (isConst || isInvMMO || isNoClobberMMO);
```
What's the logic behind the `isConst || !isVolatileMMO` part of this predicate? const && volatile doesn't make much sense to me, so why isn't this just `!isVolatileMMO`?

petar-avramovic replied:
It is copied from current implementation in AMDGPURegisterBankInfo::isScalarLoadLegal
isConst checks for address space in MMO, which can be different that address space of pointer operand

nhaehnle reviewed (commented):
I have a bunch of comments, but apart from that the change LGTM

nhaehnle reviewed (approved):
LGTM

petar-avramovic committed:
AMDGPU/GlobalISel: RegBankLegalize rules for load

Add IDs for bit width that cover multiple LLTs: B32 B64 etc.
"Predicate" wrapper class for bool predicate functions used to
write pretty rules. Predicates can be combined using &&, || and !.
Lowering for splitting and widening loads.
Write rules for loads to not change existing mir tests from old
regbankselect.

petar-avramovic commented:
### Merge activity

* **Jan 24, 6:34 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/112882).
